File: 20250418/leetcode2364. 统计坏数对的数目/main.cpp
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  long long countBadPairs(vector<int> &nums) {
    long long n = nums.size();
    long long goodPairs = 0;
    // 总数对
    long long totalPairs = ((n - 1) * n) / 2;
    unordered_map<int, int> map;
    for (int i = 0; i < n; i++) {
      // 计算当前元素的差值
      int diff = nums[i] - i;
      if (map.count(diff)) {
        goodPairs += map[diff];
      }
      // 统计差值出现的次数
      map[diff]++;
    }
    return totalPairs - goodPairs;
  }
};

File: 20250418/leetcode2364. 统计坏数对的数目/main.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var countBadPairs = function (nums) {
  let n = nums.length;
  let totalPairs = (n * (n - 1)) / 2; // 总的数对数量
  let goodPairs = 0;
  let map = new Map();

  for (let i = 0; i < n; i++) {
    let diff = nums[i] - i;
    if (map.has(diff)) {
      goodPairs += map.get(diff);
      map.set(diff, map.get(diff) + 1);
    } else {
      map.set(diff, 1);
    }
  }

  return totalPairs - goodPairs; // 坏数对 = 总数对 - 好数对
};

/**
 * @param {number[]} nums
 * @return {number}
 */
var countBadPairs = function (nums) {
  // 好数对：nums[j] - nums[i] == j - i  => nums[j] - j == nums[i] - i
  const n = nums.length;
  // i < j的总对数为 n - 1 + n - 2 + ... + 1 = n * (n - 1) / 2
  const totalPairs = (n * (n - 1)) / 2;
  let goodPairs = 0; // 计算好数对的个数
  const hash = new Map(); // 遍历j时，记录前面i的值的个数
  for (let j = 0; j < n; j++) {
    if (hash.has(nums[j] - j)) goodPairs += hash.get(nums[j] - j);
    // 记录i的值的个数
    hash.set(nums[j] - j, (hash.get(nums[j] - j) || 0) + 1);
  }
  return totalPairs - goodPairs; // 坏数对 = 总数对 - 好数对
};
